File: simulation/OathAndCoin.Simulation/Decisions/Actions.cs
using System.Collections.Immutable;
using OathAndCoin.Simulation.Ids;

namespace OathAndCoin.Simulation.Decisions;

/// <summary>
/// The fixed action vocabulary a hero's decision selects from
/// (<see cref="DecisionResult.SelectedAction"/>,
/// <see cref="DecisionResult.ConsideredActions"/>). These are actions, not
/// targets — which hero and which contract a decision concerned is carried
/// by the <see cref="Events.DomainEvent"/> the decision produces, not by the
/// action or by <see cref="CausalTrace"/> (see the remarks on
/// <see cref="CausalTrace"/>).
/// </summary>
/// <remarks>
/// Declared here as named constants — not built inline via
/// <c>ContentId.Parse("action:accept")</c> at each call site — for the same
/// reason <see cref="ReasonCodes"/> is: a value assembled ad hoc at the call
/// site can drift from every other call site that assembles "the same"
/// value independently.
///
/// <c>action:accept</c> and <c>action:decline</c> are fixed engine actions,
/// never content: no content pack defines an <c>action:</c>-namespaced
/// <see cref="ContentId"/>, and a content loader/validator must not attempt
/// to resolve these against loaded content — they do not need to exist
/// there, by design.
/// </remarks>
public static class Actions
{
    public static readonly ContentId Accept = ContentId.Parse("action:accept");
    public static readonly ContentId Decline = ContentId.Parse("action:decline");

    /// <summary>
    /// Both actions above, in declaration order — the same service
    /// <see cref="ReasonCodes.All"/> performs for its own vocabulary, and for
    /// the same reason: <c>OathAndCoin.Presentation.ActionKeys</c> needs every
    /// action this engine can select in order to check that the locale
    /// catalogue names each one, and a list written out a second time there
    /// drifts from this one the day a third action is added.
    /// <c>ActionsTests.All_ListsEveryPublicAction</c> holds this list to the
    /// class's own fields by reflection, so it cannot fall behind them either.
    /// </summary>
    public static readonly ImmutableArray<ContentId> All = ImmutableArray.Create(Accept, Decline);
}

File: packages/simulation/src/decisions/actions.ts
import { parseContentId, type ContentId } from '../ids/content-id.ts';

/**
 * The fixed action vocabulary a hero's decision selects from
 * (`DecisionResult.selectedAction`, `DecisionResult.consideredActions`). These are
 * actions, not targets — which hero and which contract a decision concerned is carried
 * by the `DomainEvent` the decision produces, not by the action or by the trace.
 *
 * Declared here as named constants — not `parseContentId('action:accept')` at each call
 * site — for the same reason `ReasonCodes` is: a value assembled ad hoc at one call site
 * drifts from the "same" value assembled independently at another.
 *
 * `action:accept` and `action:decline` are fixed engine actions, never content: no
 * content pack defines an `action:`-namespaced identifier, and a content loader must not
 * try to resolve these against loaded content — by design they do not exist there.
 */

export const Actions = Object.freeze({
  Accept: parseContentId('action:accept'),
  Decline: parseContentId('action:decline')
});

/**
 * Both actions above, in declaration order — the service `REASON_CODES` performs for its
 * own vocabulary, and derived the same way so that a third action cannot exist without
 * being in this list. The presentation layer needs every action the engine can select in
 * order to check that the locale catalogue names each one.
 */
export const ACTIONS: readonly ContentId[] = Object.freeze(Object.values(Actions));
